src/core/bittorrent/torrenthandle.cpp

```diff
@@ -1560,6 +1562,23 @@ void TorrentHandle::handleFileRenamedAlert(libtorrent::file_renamed_alert *p)
         // Renaming a file corresponds to changing the save path
         m_session->handleTorrentSavePathChanged(this);
     }
+
+    --m_renameCount;
+    if (!isMoveInProgress() && m_renameCount == 0) {
+        while (!m_torrentFinishedTriggers.isEmpty())
+            m_torrentFinishedTriggers.takeFirst()();
+    }
+}
+
+void TorrentHandle::handleFileRenameFailedAlert(libtorrent::file_rename_failed_alert *p)
+{
+    Q_UNUSED(p);
+
+    --m_renameCount;
+    if (!isMoveInProgress() && m_renameCount == 0) {
+        while (!m_torrentFinishedTriggers.isEmpty())
+            m_torrentFinishedTriggers.takeFirst()();
```

Comment:
>It's caused by m_moveFinishedTriggers.append(boost::bind(&TorrentHandle::adjustActualSavePath_impl, this)); right?

In particular, Yes.
But in General, if it is not prohibited to move/rename files inside the trigger, we should do it the way I described, even if we don't have any such trigger.